### DISCUSSION TITLE
usage with vitejs

### DISCUSSION THREAD

**Discussion by nichita-pasecinic** [runtime-env/import-meta-env#1] [Q&A]
Hi, how does it work with projects that are already using vite.js as bundler ?
I've install the dependency and run the `vite build`, but seems that my `import.meta.env` was totally handled by vitejs and they were not replaced with placeholder so I can use the CLI afterwards.

Could you please help me ?

**Comment by nichita-pasecinic** [comment]
I guess I've found the plugin in docs )))

**Reply by soc221b** [reply]
Ok, if you run into any problems, just let me know. 😃

**Reply by nichita-pasecinic** [reply]
@iendeavor, a small question, is the `.env.example.public` really required and if so, why we cant we use its' values as the default dev env (instead of defining a new `.env`) ?

**Reply by nichita-pasecinic** [reply]
in examples there is a specific example with vite config with different build name out directory, meaning that we need to pass some different options to CLI in order to work with the envs. I've changed the build directory name from vitejs and seems like it's still working without any need to pass the out-dir name as CLI arguments, this will just work - `import-meta-env --example .env.example`
Are there any considerations about this ?

**Reply by soc221b** [reply]
@nichita-pasecinic 

> is the `.env.example.public` really required

Yes, we need to explicitly define which environment variables need to be used.

Without this, we could accidentally expose all environment variables, and cause unexpected leaking of sensitive information.

> if so, why we cant we use its' values as the default dev env (instead of defining a new `.env`) ?

There are some reasons:

1. Each developer/project may have its own environment variables (monorepo).
2. No need to worry that someone might accidentally commit the `.env.example.public` file. Instead, the `.env` will be gitignored and we can change it at will (e.g., switch to production to track issues).
3. Keep things simple. In current version <sup>1</sup>, your environment variables can only be injected from a `.env` file and/or from the system (that's why we don't support multiple files like `.env.development`, `.env.production`).

---

<sup>1</sup> current versions are `@import-meta-env/babel@0.1.9`, `@import-meta-env/cli@0.1.7`, and `@import-meta-env/unplugin@0.1.8`

**Reply by soc221b** [reply]
> in examples there is a specific example with vite config with different build name out directory, meaning that we need to pass some different options to CLI in order to work with the envs. I've changed the build directory name from vitejs and seems like it's still working without any need to pass the out-dir name as CLI arguments, this will just work - `import-meta-env --example .env.example` Are there any considerations about this ?

The current CLI plugin detects these common output directories by default:

https://github.com/iendeavor/import-meta-env/blob/74701b221e4a46e5617c4f5701f930bfe0e5fab9/packages/cli/src/shared.ts#L5-L16

Maybe your `out-dir` is one of the default directories 😂?

You can also [run CLI help](https://iendeavor.github.io/import-meta-env/api.html#import-meta-env-cli) to get this information.

**Reply by nichita-pasecinic** [reply]
@iendeavor thanks for a answering my questions, now I got it 😄, well yeah, the out-dir indeed it's a `build`, though using it with vitejs, but if it is listed here under defaults for cra, then I can't wish any more )))

**Reply by nichita-pasecinic** [reply]
@iendeavor one last question ))

I've tried to push my changes to serve the build with the approach that this library provides from an docker container and I have a following error on permissions:

```
> import-meta-env -x .env.public && serve -s -S build

node:internal/fs/utils:344
    throw err;
    ^

Error: EACCES: permission denied, copyfile '/usr/src/app/build/assets/import-meta-env.8fdefd1c.js' -> '/usr/src/app/build/assets/import-meta-env.8fdefd1c.js.bak'
    at copyFileSync (node:fs:2804:3)
    at /home/node/lib/node_modules/@import-meta-env/cli/bin/import-meta-env.js:4457:35
    at Array.forEach (<anonymous>)
    at main (/home/node/lib/node_modules/@import-meta-env/cli/bin/import-meta-env.js:4443:34)
    at Object.<anonymous> (/home/node/lib/node_modules/@import-meta-env/cli/bin/import-meta-env.js:4463:3)
    at Module._compile (node:internal/modules/cjs/loader:1103:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1157:10)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:77:12) {
  errno: -13,
  syscall: 'copyfile',
  code: 'EACCES',
  path: '/usr/src/app/build/assets/import-meta-env.8fdefd1c.js',
  dest: '/usr/src/app/build/assets/import-meta-env.8fdefd1c.js.bak'
}
```
mb do you know what is about and it could be fixed ?
The only think I understand here is that the cli does not have enought permissions to do a backup for the `import-meta-env.8fdefd1c.js` file.

Thank you in advance!

**Reply by soc221b** [reply]
@nichita-pasecinic

It can be solved by giving it sufficient permissions using `chown` and/or `chmod`, you can read more about it here:
https://stackoverflow.com/questions/48910876/error-eacces-permission-denied-access-usr-local-lib-node-modules
https://stackoverflow.com/questions/3740152/how-to-change-permissions-for-a-folder-and-its-subfolders-files-in-one-step

If you don't need to restore/backup feature, you can pass `--disposable` to skip backup process:
https://github.com/iendeavor/import-meta-env/blob/main/packages/cli/src/create-command.ts#L32-L35

Hope this helps you. 😃

**Reply by nichita-pasecinic** [reply]
Thanks legend ))
issue solved by 
```
RUN chown -R node:node /usr/src/app
```
